clusterloader2/pkg/measurement/common/kcm_metrics.go

```diff
@@ -0,0 +1,223 @@
+		stopCh:       make(chan struct{}),
+	}
+}
+
+type kcmWorkQueueDepthMeasurement struct {
+	// depthMetrics is the minimum workqueue depth for each resouce in the leader.
+	depthMetrics      map[string]float64
+	hpaAllowedBacklog *float64
+
+	isRunning bool
+	stopCh    chan struct{}
+}
+
+func (m *kcmWorkQueueDepthMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	provider := config.ClusterFramework.GetClusterConfig().Provider
+	hosts := config.ClusterFramework.GetClusterConfig().MasterIPs
+	SSHToMasterSupported := provider.Features().SupportSSHToMaster
+
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	kcmPort := config.ClusterFramework.GetClusterConfig().KCMPort
+	switch action {
+	case "start":
+		klog.V(2).Infof("%s: starting kcm metrics collecting...", m)
+		if !SSHToMasterSupported {
+			klog.Warningf("ssh to master node is not supported for provider: %v", provider.Name())
+			return nil, nil
+		}
+		m.start(provider, hosts, kcmPort)
+		return nil, nil
+	case "gather":
+		m.Dispose()
+		hpaAllowedBacklog, err := util.GetFloat64OrDefault(config.Params, "hpaAllowedBacklog", -1)
+		if err != nil {
+			return nil, err
+		}
+		if hpaAllowedBacklog >= 0 {
+			m.hpaAllowedBacklog = &hpaAllowedBacklog
+		}
+		summary, err := m.createSummary()
+		return []measurement.Summary{summary}, err
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+// Dispose cleans up after the measurement.
+func (m *kcmWorkQueueDepthMeasurement) Dispose() {
+	if m.isRunning {
+		m.isRunning = false
+		close(m.stopCh)
+	}
+}
+
+// String returns string representation of this measurement.
+func (m *kcmWorkQueueDepthMeasurement) String() string {
+	return kcmWorkQueueDepthMetricName
+}
+
+func (m *kcmWorkQueueDepthMeasurement) start(provider provider.Provider, hosts []string, port int) error {
+	if m.isRunning {
+		return fmt.Errorf("%s: measurement already running", m)
+	}
+	m.isRunning = true
+	klog.V(3).Infof("collecting KCM workqueue depth metrics from hosts: %v", hosts)
+	if len(hosts) == 0 {
+		return nil
+	}
+
+	// There's only one goroutine that collect metrics from all KCM instances.
+	go func() {
+		for {
+			select {
+			case <-time.After(time.Second):
+				metricsFromHosts := []map[string]float64{}
+				for _, host := range hosts {
+					workQueueDepthMetrics, err := m.getKCMWorkQueueDepth(host, provider, port)
+					if err != nil {
+						klog.Errorf("%s: failed to collect KCM workqueue depth: %v", m, err)
+						continue
+					}
+					metricsFromHosts = append(metricsFromHosts, workQueueDepthMetrics)
+				}
+				if len(metricsFromHosts) == 0 {
+					continue
+				}
+				// For one single data point in each KCM instance, we take the max,
+				// since the leader will have larger workqueue backlog.
+				result := metricsFromHosts[0]
+				for i := 1; i < len(metricsFromHosts); i++ {
+					result = getMaxOrMin(result, metricsFromHosts[i], true)
```

Comment:
Do you think you are getting the `max` here ? Because the `getMaxOrMin` function doesn't seem to take the boolean into consideration. So i think you might be getting always the `min` here IIUC ?